### PR TITLE
Ensure `BaseCoordinateFrame` separation methods work with `SkyCoord` input

### PR DESCRIPTION
**Pull request by eerovaher**
### Description

The first commit here adds a test which demonstrates that it is possible to get wrong results from `BaseCoordinateFrame` `separation()` and `separation_3d()` methods if the input to those methods is a `SkyCoord` instance. These methods do not claim to support `SkyCoord` input, but if they don't work with `SkyCoord` they should fail in an obvious manner. The current behavior is therefore a bug and the second commit here fixes it.

~The third commit removes the `SkyCoord` `separation()` and `separation_3d()` methods. This can be removed without loss of functionality after fixing the bug because a `SkyCoord` instance then exposes the separation methods of its underlying frame.~
EDIT: there will be a separate pull request for the third commit.

- [x] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/coordinates/sky_coordinate.py [hunk 1]
```diff
@@ -1132,6 +1132,18 @@ def is_equivalent_frame(self, other):
                 "Tried to do is_equivalent_frame on something that isn't frame-like"
             )
 
+    def _prepare_for_separation(self, other):
+        if not self.is_equivalent_frame(other):
+            kwargs = {"merge_attributes": False} if isinstance(other, SkyCoord) else {}
+            try:
+                other = other.transform_to(self, **kwargs)
+            except TypeError:
+                raise TypeError(
```
Comment (neutrinoceros) [inline astropy/coordinates/sky_coordinate.py:1141]: while we're at it maybe we could also chain these exceptions either as

```python
try:
    ...
except TypeError as exc:
    raise TypeError(...) from exc
```
or
```python
try:
    ...
except TypeError:
    raise TypeError(...) from None
```
? (I think the second one is more appropriate here)

Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1141]: It doesn't look like this code path can ever be triggered. Trying to compute the separation of a frame without data raises a `ValueError`, not a `TypeError`:
```python
>>> from astropy.coordinates import FK5, SkyCoord
>>> SkyCoord(1, 1, unit="deg").separation(FK5())
Traceback (most recent call last):
  ...
ValueError: Cannot transform a frame with no data
```
To me it seems that this error message is clear enough, so this `try-except` block can be removed.

File: astropy/coordinates/sky_coordinate.py [hunk 2]
```diff
@@ -1132,6 +1132,10 @@ def is_equivalent_frame(self, other):
                 "Tried to do is_equivalent_frame on something that isn't frame-like"
             )
 
+    def _prepare_for_separation(self, other):
```
Comment (mhvk) [inline astropy/coordinates/sky_coordinate.py:1135]: How about calling this `_to_own_frame(self, other)`? I think that is a little more descriptive and possibly we need it elsewhere (e.g., in `position_angle` it looks like we're doing something similar). It is also nicely analogous with `Quantity._to_own_unit`). 

That said, I'm actually not sure this shouldn't live on `baseframe`... See general comment.

Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py:1135]: Falling back to `BaseCoordinateFrame.separation()` doesn't always work:
```python
>>> from astropy.coordinates import SkyCoord
>>> sc1 = SkyCoord(1, 2, unit="deg", frame="fk5")
>>> sc2 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="B1950")
>>> sc1.separation(sc2)
<Angle 0.69815121 deg>
>>> sc1.frame.separation(sc2)
<Angle 0. deg>
```

See also https://github.com/astropy/astropy/blob/6b36e53c2608c43a35a5561122688fb9e04879bf/astropy/coordinates/tests/test_sky_coord.py#L892

But the suggestion about the name of the new private method is a good one.

Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py:1135]: OK, slightly weird that `frame.separation` ignores `equinox` but let that be for another time. For now, good to know that we need to override.

**Comment by mhvk** [comment]
p.s. Somewhat surprising, the increase in performance. But nice!

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (ayshih) [reply astropy/coordinates/sky_coordinate.py]: Hmm, the reason for the difference is because a `SkyCoord` is being passed to `BaseCoordinateFrame.separation()`.  `SkyCoord.separation()` explicitly blocks the merging-attributes functionality of `SkyCoord.transform_to()`, but `BaseCoordinateFrame.separation()` unsurprisingly does not have such a block since `BaseCoordinateFrame.transform_to()` simply has no merging-attributes functionality.  So, when a `SkyCoord` is passed to `BaseCoordinateFrame.separation()`, the "wrong" `transform_to()` method gets called, frame attributes are merged, and `sc1` is treated as if it has the same equinox as `sc2` because `sc1`'s equinox has not been set.  If you explicitly set the equinox for `sc1`, the answer is correct:
```python
>>> from astropy.coordinates import SkyCoord
>>> sc1 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="J2000")
>>> sc2 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="B1950")
>>> sc1.separation(sc2)
<Angle 0.69815121 deg>
>>> sc1.frame.separation(sc2)
<Angle 0.69815121 deg>
```
```

Comment (ayshih) [reply astropy/coordinates/sky_coordinate.py]: This is the problematic line:
https://github.com/astropy/astropy/blob/6b36e53c2608c43a35a5561122688fb9e04879bf/astropy/coordinates/baseframe.py#L1801

**Comment by neutrinoceros** [comment]
is there something wrong with codecov reports lately or should a test or two be added to cover these (surprisingly) uncovered paths ?

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py]: That seems like a good solution; I don't think it was ever meant to be the case that a `SkyCoord` with no explicit equinox set should be interpreted as one that can have arbitrary equinox.

**Comment by pllim** [comment]
There was an error 404 in the upload step in the coverage job here but somehow codecov does not fail the job. When you see weird numbers, go into the log and check the upload step. This happens on and off because pull_request does not get the token to guarantee success. If codecov thinks we're spamming them, this happens.

Unless coverage check is critical, I usually ignore. Otherwise, maintainer can opt to rerun just that job.

The non-zero number on failure is because CircleCI also uploads coverage but only for visualization tests.

Hope this clarifies the matter!

File: astropy/coordinates/sky_coordinate.py (hunk 2 above)
Comment (eerovaher) [reply astropy/coordinates/sky_coordinate.py]: Following the suggestion of @ayshih it is indeed possible to remove the `SkyCoord` separation methods entirely without any loss of functionality. However, in that case these methods will not appear in the `SkyCoord` API documentation page. I would say that's an acceptable tradeoff, but I'd like this to be confirmed before I push the new commit. Note that in the narrative docs it is possible to replace links to the `SkyCoord` methods with links to the `BaseCoordinateFrame` methods, so there's no problem there.

Comment (mhvk) [reply astropy/coordinates/sky_coordinate.py]: I'm OK with that - it seems important that the methods behave the same way, and the easiest way to ensure that is the case is to have just one!

**Comment by eerovaher** [comment]
The changes this pull request now makes to `BaseCoordinateFrame` allow its separation methods to handle `SkyCoord` instances as arguments, whereas previously that was not the case. That might require a change log entry. I'll leave it to reviewers to remove the `no-changelog-entry-needed` label if they agree.

File: astropy/coordinates/baseframe.py [hunk 3]
```diff
@@ -1798,8 +1801,11 @@ def separation(self, other):
         from .angles import Angle, angular_separation
 
         self_unit_sph = self.represent_as(r.UnitSphericalRepresentation)
-        other_transformed = other.transform_to(self)
-        other_unit_sph = other_transformed.represent_as(r.UnitSphericalRepresentation)
+        other_unit_sph = (
+            getattr(other, "frame", other)
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py:1805]: It may still be worthwhile to define a simple `_to_own_frame` so that these pieces of code are kept together. E.g., possibly one should just do `other.frame if isinstance(other, SkyCoord) else other` or `getattr(other, "_sky_coord_frame", other)` [as is done in `transform_to`]. But, really, also happy to just have it like this.

**Review by mhvk — COMMENTED** [review]
There's part of me that worries about omitting all the `SkyCoord` non-frame attributes, but since that was the case already, clearly it is not something that is obviously a problem (and it is clear that just merging attributes *is* a problem).

I do think we should have a changelog entry, since really the old behaviour for Baseframe was a bug for `SkyCoord` input.

File: astropy/coordinates/baseframe.py (hunk 3 above)
Comment (mhvk) [reply astropy/coordinates/baseframe.py:1805]: OK, let's keep it as is here!

**Review by mhvk — APPROVED** [review]
To me, this looks all good and since there is no actual change in `SkyCoord.separation*` - from its perspective, this is just a refactoring - it seems fine to me to merge as is, including the backport. But I won't do that yet just in case others disagree; I'm fine with not backporting or splitting as well.

**Comment by eerovaher** [comment]
@pllim, could you advise if the refactoring commit should be included here, in which case it will be backported with the bugfix, or if a separate pull request should be opened for the refactoring?

**Comment by pllim** [comment]
We generally do not backport refactoring. But if it is just a few lines and totally backward compatible, *maybe* ok.

tl;dr -- new PR would be best

**Review by mhvk — APPROVED** [review]
OK, thanks for splitting the bug off! Looks all OK.

**Comment by eerovaher** [comment]
@meeseeksdev backport to v6.0.x
